src/haz3lcore/pretty/Abbreviate.re

```diff
@@ -0,0 +1,841 @@
+/* Abbreviate a term for display, specifically for the live
+ * value probe projector. This is currently specialized for
+ * expressions which are (at least partially) values.
+ *
+ * This is an ugly rough approach, and should be rewritten when
+ * we have projectors (in particular, fold) within value displays.
+ *
+ * This approach ends up duplicating way too much info
+ * with ExpToSeg. This should probably be rewritten to
+ * use that somehow. */
```

Comment:
An approach on the segment structure with abbreviations would probably work. If we want to keep this returning the `Exp` structure and use `Invalid` for abbreviations I'm not sure how it would work.

Reply:
yeah i'm not happy at all with this module; i think doing this properly is a job in itself (in principle i want this to be more flexible and not just use a 'left-to-right greedy' strategy for abbreviation; i want to be able to do things like allocate half the size budget to each side of a pair for example). my tentative goal is to redo this using projectors at some point, using a more-flexible version of the fold projector to do what I'm doing with Invalids right now. but currently this would cross abstraction boundaries in an awkward way

Reply:
That makes sense. If there's an intermediate state that operating on segments would be useful I could look at it but it's not clear if that helps get towards the projector folding version.